File: Cpp_programs/Sorting,Recursion/mergesort.cpp
#include<iostream>
using namespace std;

void merge(int arr[],int s,int e)
{

    int mid=s+(e-s)/2;
    int len1=mid-s+1;
    int len2=e-mid;

    int *first = new int[len1];
    int *second = new int[len2];

    //copy values
    int k=s;
    for(int i=0;i<len1;i++)
    first[i]=arr[k++];

    k=mid+1;
    for(int i=0;i<len2;i++)
    second[i]=arr[k++];

    //merge 2 sorted arrays
    int i=0,j=0;
    k=s;
    while(i<len1 && j<len2)
    {
        if(first[i] < second[j])
        arr[k++]=first[i++];
        else
        arr[k++]=second[j++];
    }
    while(i<len1)
    arr[k++]=first[i++];
    while(j<len2)
    arr[k++]=second[j++];

    delete[] first;
    delete []second;

}


void mergesort(int arr[],int s,int e)
{
    //base case
    if(s>=e)
        return ;

    int mid=s+(e-s)/2;

    mergesort(arr,s,mid);
    mergesort(arr,mid+1,e);

    merge(arr,s,e);
}


int main()
{
    int arr[5]={5,4,3,2,1};
    int n=5;

    mergesort(arr,0,n-1);

    for(int i=0;i<n;i++)
    cout<<arr[i]<<" ";
    cout<<endl;
}

File: Python_programs/Linked_list,Sliding_window/mergesort.py
l1=l[start:mid+1] 
l2=l[mid+1:end+1] 
i=0 
j=0 
k=start 
while i<len(l1) and j<len(l2): 
if l1[i]<l2[j]: 
l[k]=l1[i] 
i+=1 
k+=1 
else: 
l[k]=l2[j] 
j+=1 
k+=1 
while i<len(l1): 
l[k]=l1[i] 
k+=1 
i+=1 
while j<len(l2): 
l[k]=l2[j] 
j+=1 
k+=1 
def merge_sort(l,start,end): 
if end>start: 
mid=(start+end)//2 
merge_sort(l,start,mid) 
merge_sort(l,mid+1,end) 
merge(l,start,mid,end) 
l=[7,0,1,4,8,2,4] 
merge_sort(l,0,len(l)-1) 
print(l)